pinc/filter_project_list.inc

```diff
@@ -11,485 +12,418 @@ include_once($relPath.'misc.inc'); // attr_safe()
+        new ProjectSearchElement("special_code", $data),
+        new ProjectSearchElement("difficulty", $data),
+        new LanguageSearchElement("language", $data),
+    ];
 
-    // split the filter into segments; each segment will contain
-    // the ( ... OR ... ) string for one form field
-    $segments = explode(" AND ", $filter);
-    foreach($segments as $segment)
+    $filter = "";
+    foreach($search_elements as $element)
     {
-        // remove trailing quotes, spaces, or parens
-        $segment = rtrim($segment,"' )");
-        $segment = ltrim($segment,"' (");
+        $filter .= $element->get_sql_component();
+    }
+    return $filter;
+}
 
-        // split each segment into a set of column='value' pairs
-        $value_pairs = explode(" OR ", $segment);
-        foreach($value_pairs as $value_pair)
-        {
-            // split the column='value' pairs into their two parts
-            $value_pair = explode("=", $value_pair);
-            if (count($value_pair) != 2) continue;
-            list($field, $value) = $value_pair;
+function sql_regex_escape($value)
+{
+    // double the backslashes introduced by quotemeta
+    return str_replace('\\', '\\\\', quotemeta($value));
+}
 
-            // clean up the $value, which looks like: 'value'
-            $value = rtrim($value,"' ");
-            $value = ltrim($value,"' ");
+function get_lang_match($data)
+{
+    // set a default if not set
+    return array_get($data, "lang-match", "primwith");
+}
 
-            // skip any empty values
-            if($value=="") continue;
+function save_data($pguser, $filter_type, $data)
+{
+    save_raw_data($pguser, "{$filter_type}_data", serialize($data));
+}
 
-            if(!is_array(@$data[$field]))
-                $data[$field]=array();
+function save_display($pguser, $filter_type, $data)
+{
+    save_raw_data($pguser, "{$filter_type}_display", $data);
+}
 
-            $data[$field][]=$value;
-        }
-    }
+function save_raw_data($pguser, $data_name, $data)
+{
+    $enc_data = mysqli_real_escape_string(DPDatabase::get_connection(), $data);
+    $query = "REPLACE INTO user_filters (username, filtertype, value) VALUES ('$pguser', '$data_name', '$enc_data')";
+    mysqli_query(DPDatabase::get_connection(), $query) or die(mysqli_error(DPDatabase::get_connection()));
+}
 
-    return $data;
+function get_saved_data($pguser, $filter_type)
+{
+    return unserialize(get_raw_saved_data($pguser, "${filter_type}_data"));
 }
 
-function _load_project_filter_field_values($field, $state_sql)
-// Given a specific field, return all possible values for that field.
+function get_project_filter_display($pguser, $filter_type)
 {
-    $load_associative = FALSE;
+    return get_raw_saved_data($pguser, "${filter_type}_display");
+}
 
-    switch($field)
+function get_raw_saved_data($pguser, $data_name)
+{
+    $query = "SELECT value FROM user_filters WHERE username = '$pguser' AND filtertype = '$data_name'";
```

Comment:
We should escape `$data_name` to double-ensure that this can't result in an SQL injection sometime in the future.

Reply:
Done this if I've understood correctly.